[HUDI-1985] Migrate the hudi site to docusaurus platform (website complete re-design)

vingov opened the pull request:
## *Tips*
- *Thank you very much for contributing to Apache Hudi.*
- *Please review https://hudi.apache.org/contributing.html before opening a pull request.*

## What is the purpose of the pull request
- Migrated the static site generator to the Docusaurus platform, to add the following features plus easy to add more features:
- Side bar, multi-level pages. Quick Start > Spark , Quick Start > Flink
- Code tab for different scala, java, python, SQL.
- Auto maximize for larger screens
- Show 0.8.0 (latest release) and keep dev separate for docs.
- Native i18n support.
- Better versioning support for docs.
- Re-arranged the content to match the [new re-design](https://www.figma.com/file/tipod1JZRw7anZRWBI6sZh/Hudi.Apache?node-id=32%3A6).
- Ability to add search on the site, after the first version is published search will be enabled.
- Old `docs` folder will be removed in the next PR.

## Brief change log

- *Migrated the static site generator to the Docusaurus platform, to add the following features plus easy to add more features.*

## Verify this pull request

*(Please pick either of the following options)*

This change added tests and can be verified as follows:

  - *Run these commands to test the site locally at [http://localhost:3000/](http://localhost:3000/):*
  Note: ensure you have npm installed
 ```bash
 ./website/scripts/build-site.sh
cd website
npm run serve
```
  - *Manually verified the website re-design changes by running the site locally.*

## Committer checklist

 - [x] Has a corresponding JIRA in PR title & commit
 
 - [x] Commit message is descriptive of the change
 
 - [x] CI is green

 - [x] Necessary doc changes done or have another open PR
       
 - [x] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA. More re-design changes will be on a new JIRA sub-task.

vingov commented:
Created a [short demo](https://www.awesomescreenshot.com/video/4661736?key=12a221a8a04e097a99200c061d32ca27) of the new site re-design with docusarus platform to make it easy for the reviewers.

codope commented:
@vingov @nsivabalan Shall we move the schema evolution subsection to a new page under documentation? It's gonna be a story of its own.

vingov commented:
> @vingov @nsivabalan Shall we move the schema evolution subsection to a new page under documentation? It's gonna be a story of its own.

Yes, we can do all structural changes after we land this version.

vingov commented:
@nsivabalan - Can you please review this PR?

vingov commented:
Summary of changes:

1. Docusaurus doesn't have the concept of permalink in the header, the URL is formed based on the folder and file name, hence renamed all the files with the permalink to preserve the old path and to avoid broken links for the docs.
2. Since the new navigation structure is different for the non-doc static pages, I've updated all the references in the document.
3. Docusaurus doesn't support a few of the Jekyll features like:

- The HTML class attribute should be called className, hence updated those.
- In some cases, the <span> tag messes up the formatting of the content hence got rid of it.
- Fixed the broken HTML tags.
- For loops got converted to a theme component for the powered-by page.
- Keywords should be an array instead of a comma-separated string.

4. Rebuilt the left sidebar and all old docs version sidebar and the main dropdown navigation.
5. Moved around a bunch of files and renamed some to support versioning and i18n support.
6. Most of the classic theme looked good already out of the box, fixed a bunch of them in the src/css/custom.css.
7. Rebuild the home page as a pure js rendered page src/pages/index.js.
8. Built the footer navigation with all relevant links to improve the SEO and ease of use and improve the visibility of the inner pages.
9. Broken the releases page to individual pages, to improve the usability and to support the left side bar.
10. Broken the community page into 3 pages:
- Get Involved.
- Team.
- How to Contribute
11. Added the code tab support for the quick-start spark guide, the same could be easily applied for other pages.

Please refer to the [docusaurus official documentation ](https://docusaurus.io/docs/markdown-features)for adding more features to the site.

This is not the final version, the first step in the process is to migrate the content as-is with few minor changes to the docusaurus platform, then we can make incremental changes to add more features.

I'll be adding the search bar in the next iteration.

These sed one-liners saved the day for me!!

```
sed -i '' -E 's/src=("[^"]*")/src={require(\1).default}/g' *.md
sed -i '' -E 's/keywords:(.*)$/keywords: [\1]/g' *.md
sed -i '' -E 's/<br>/<br\/>/g' *.md
sed -i '' -E 's/style=("[^"]*")//g' *.md
sed -i '' -E 's/class=("[^"]*")/className=\1/g' *.md
sed -i '' -E 's|/docs/0.5.0-|/docs/|g' *.md
sed -i '' -E 's|/docs/0.5.1-|/docs/|g' *.md
sed -i '' -E 's|/docs/0.5.2-|/docs/|g' *.md
sed -i '' -E 's|/docs/0.5.3-|/docs/|g' *.md
sed -i '' -E 's|/docs/0.6.0-|/docs/|g' *.md
sed -i '' -E 's|/docs/0.7.0-|/docs/|g' *.md
sed -i '' -E 's|/docs/0.8.0-|/docs/|g' *.md

find . -name '*.md' -exec sed -i '' '/permalink:/d' {} +

for filename in *.cn.md; do mv $filename ${filename//cn.md/md}; done
```

nsivabalan reviewed (commented):
I built it locally and verified all the links to the best of my knowledge. Things are looking ok. 
Few minor comments. Some of them are for follow ups, don't necessarily need to address it in this patch. 
1. In the Landing page towards the bottom -> About -> roadmap: Guess this is bit outdated. May be we can link to the latest blog. or remove it.
2. On the top right, where we can click to view all different versions, I see "Next" along with other versions. Not sure if it was added by mistake. 
3. I don't see Robinhood in powered by. May we should add it. 
4. In the configurations page, if we can tabify spark datasource configs and flink configs, would be nice. But this tabs will span only partially and not full page. Bcoz, writeclient configs are same for both. 

Overall, good job!

vingov commented:
Thanks for the review @nsivabalan!

1. Updated the roadmap link to streaming data lake's latest blog link.
2. The "Next" link shows the latest master branch docs, Next meaning the next unreleased version docs, it's not selected by default, only the latest stable release docs are shown by default, the next version is easier for developers to validate their docs changes. - There is a flag to hide the next link, if you still prefer to hide it, let me know I can disable it.
3. Added Robinhood to the powered-by page.
4. Yes, we can make use of the tabs to separate out the spark/flink configs, I see that there are few changes already in the latest unreleased version, we can take that later.

vingov commented:
Moreover, for the next version, there is a banner added to let users know that these docs are for the unreleased version:

<img width="917" alt="Screen Shot 2021-08-01 at 2 06 24 PM" src="https://user-images.githubusercontent.com/1142498/127785355-e39f3aa4-1011-4259-83a1-9a806e687839.png">

vinothchandar commented:
@nsivabalan lets keep this to just reflecting the current site as-is and avoid adding new links or powered-by? We can do these in smaller steps after? 

@vingov Once we land this, I assume travis will be broken and we will have to fix the automatic site build in a follow on PR? Just trying to understand the plans

vingov commented:
@vinothchandar - I have fixed the Travis script as well, similar to the jekyll script, it will build and move it to content folder and it will push it back to asf-site.

vinothchandar commented:
@vingov this is awesome!

vinothchandar commented:
- [ ] Use `Current` instead of `Next` for denoting the master build
- [ ] Retain the hyperlinks from the current site's footer

![image](https://user-images.githubusercontent.com/1179324/127908550-9b524b8e-45e7-47da-9c16-d60f1cf07fc7.png)

- [ ] Call the blog "our vision" instead of "streaming data lake"

![image](https://user-images.githubusercontent.com/1179324/127908744-08db8cd8-624d-4b29-9222-e4fbbae81c97.png)
 
- [ ] Twitter link to point to `/apachehudi` without camel casing
- [ ] Extra "Concepts" link under Resources. Remove `comparison` , migration etc. just match this to current site?

Looks good to me overall. I would just check all changes to docs in the last 1 month again just to be sure. we got evertying

vinothchandar commented:
@vingov this is really good work!  and also very laborious. Thanks for doing this! Once we address these, we can land this

vingov commented:
Thanks for the feedback @vinothchandar!

1. Updated the Next link to Current.
2. Retained the apache links on the footer, but those links are already in the Apache footer menu as well, as far as the [ASF rules](https://www.apache.org/foundation/marks/pmcs#navigation), we should be good without those links as well.
3. Renamed the blog link to "Our Vision"
4. When I checked my local site, the Twitter links were using proper camel casing both on the header and footer, can you please check again?
`$ grep -n "twitter.com/ApacheHudi" docusaurus.config.js
176:          href: 'https://twitter.com/ApacheHudi',
296:              href: 'https://twitter.com/ApacheHudi',` 
5. Removed these extra links on resources, to match the current site:
- concepts
- comparison
- migration

PTAL!
